src/collection/ArrayList.php

```diff
@@ -93,12 +119,65 @@ public function remove($element) {
 	 * @param array|Iterator $collection
 	 * @return $this
 	 */
-	public function removeAll($collection) {
+	public function removeAll($collection): self {
 		foreach ($collection as $element) {
 			$this->remove($element);
 		}
 		
 		return $this;
 	}
 
+	/**
+	 * Searches the collection with a given callback and returns the index for the first element if found.
+	 *
+	 * The callback function takes one or two parameters:
+	 *
+	 *     function ($element [, $query]) {}
+	 *
+	 * The callback must return a boolean
+	 *
+	 * @param mixed $query OPTIONAL the provided query
+	 * @param callable $callback the callback function
+	 * @return int|null the index or null if it hasn't been found
+	 */
+	public function findIndex(): ?int {
```

Comment:
What's the matter of moving the index finders to `ArrayList` instead of keeping them in `AbstractList`?

Reply:
`findIndex` uses `indexOf` method which is implemented in `ArrayList` only. Index finders are "generic" methods and their natural place is a "parent" class (`AbstractList`) but actually they're used in `ArrayList` only: neither in `Set`, nor in `Stack`, nor in `Queue`, which don't have and `indexOf` implementation.

If we don't want to move the method, maybe we could define an abstract method `indexOf` in `AbstractList`, but in this case we force ourselves to implement an `indexOf` method in `Set`, `Stack` and `Queue` that really don't care about indexes.